### PR TITLE
adapt to gFTL v2 container

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
define containers by the way of gFTL-v2 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [ ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -52,6 +53,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Major refactoring of GenericSetServices
+  Work is not completed, but a new layer is introduced with the intent that the user SetServices is called
+  from with in the new layer as opposed to the previous mechanism that obligated user SetServices to call
+  generic.   That call is now deprecated.   Significant cleanup remains.
```
Comment (tclune) [inline CHANGELOG.md:59]: @mathomp4 I think this must be from a garbled history.  Isn't this only on the v3 branch?

**Review by tclune — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (mathomp4) [reply CHANGELOG.md:59]: Yeah. That shouldn't be on `develop` anymore. It's almost like Weiyuan started before we did the Great Reversion, but GitHub played it "safe" during a merge?

For now he should just remove this entry and push

Comment (tclune) [reply CHANGELOG.md:59]:
```suggestion.
```
Making it easy for @weiyuan-jiang  - just has to click the button.

**Comment by weiyuan-jiang** [comment]
> 

Yes, this MAPL needs new version of gFTL

File: CHANGELOG.md (hunk 1 above)
Comment (weiyuan-jiang) [reply CHANGELOG.md:59]: > 

Exactly, I am puzzled when I tried to resolve the conflict

Comment (weiyuan-jiang) [reply CHANGELOG.md:59]: > 

Why I can't see the button?

Comment (tclune) [reply CHANGELOG.md:59]: I think it because I used the "suggest" button after this thread was started.  Let me try a new review.

**Review by tclune — APPROVED** [review]

**Comment by mathomp4** [comment]
Okay. Let me try and get the new Baselibs spun up. This will take a bit of time as I'll need to roll it out everywhere and fix up GEOSgcm.

**Comment by weiyuan-jiang** [comment]
> Okay. Let me try and get the new Baselibs spun up. This will take a bit of time as I'll need to roll it out everywhere and fix up GEOSgcm.

This PR will not work with GEOSgcm either. After I have new baselib, I will need to change GEOSgcm too.

**Comment by mathomp4** [comment]
> > Okay. Let me try and get the new Baselibs spun up. This will take a bit of time as I'll need to roll it out everywhere and fix up GEOSgcm.
> 
> This PR will not work with GEOSgcm either. After I have new baselib, I will need to change GEOSgcm too.

oooh. Really? Okay. Is this needed for GOCART2G stuff?

**Comment by mathomp4** [comment]
> No - this is not for GOCART2G. Just a mini modernization effort that will allow us to use some new features of the v2 containers.
> 
> I probably misstated things in a separate thread to @weiyuan-jiang . GEOSgcm can only use the new MAPL if it also updates the baselibs. So the order of releases is
> 
> 1. baselibs
> 2. MAPL
> 3. esma_env
> 4. GEOSgcm (and other fixtures)
> 
> I'm assuming upticks on baselibs are not a big problem for GEOSgcm.

Ohh. So that means we don't require changes to GEOSgcm for this MAPL? Or do we need to fix any/all gFTL code in the model?

**Comment by tclune** [comment]
No - this is a backward compatible change other than the dependency on the fixes in gFTL.

**Comment by mathomp4** [comment]
Okay. Once I get the new images spun up, I'll update this branch. @weiyuan-jiang Fair warning that'll probably bring in `develop` as I'll probably need to press the "Update branch" button.

**Comment by weiyuan-jiang** [comment]
> 

It is not backward compatible when GEOSgcm uses the containers defined in MAPL. Those get(), value(), key() should be changed accordingly

**Comment by tclune** [comment]
It _is_ backward compatible.  The V1 interfaces are still available and those gridcomps can still use them.   Yes - it would be a good idea to update them, but these changes have nothing to do with that.

**Comment by mathomp4** [comment]
Okay. I think I've updated the images. We'll see how the CI goes. 

If all looks good, I'll then work on making a new ESMA_env version as well.

**Comment by mathomp4** [comment]
This is fun. I think this PR is now broken by changes @bena-nasa made in #1400 😄

**Comment by weiyuan-jiang** [comment]
> This is fun. I think this PR is now broken by changes @bena-nasa made in #1400 😄

I fixed this

**Comment by mathomp4** [comment]
Marking as long-term.

**Comment by mathomp4** [comment]
@weiyuan-jiang Some notes. The CI is now failing with (for example):

```
[ 39%] Building Fortran object pfio/CMakeFiles/MAPL.pfio.dir/FileMetadata.F90.o
/root/project/MAPL/pfio/FileMetadata.F90(251): error #8010: A pointer dummy argument with the INTENT(IN) attribute shall not appear as an pointer-object in a pointer-assignment-stmt or nullify-stmt.   [THIS]
      var => this%variables%at(var_name)
-------------^
compilation aborted for /root/project/MAPL/pfio/FileMetadata.F90 (code 1)
```

I talked with @tclune and he had a few thoughts:

1. `this` has to be `intent(inout)`, but I think @tclune is not the happiest with this solution
2. Maybe try making a temp variable to do this with, though @tclune thinks the compiler will see through it
3. @tclune rewrites bits of gFTL 2.

We should let @tclune comment more

**Comment by weiyuan-jiang** [comment]
> @weiyuan-jiang Some notes. The CI is now failing with (for example):
> 
> ```
> [ 39%] Building Fortran object pfio/CMakeFiles/MAPL.pfio.dir/FileMetadata.F90.o
> /root/project/MAPL/pfio/FileMetadata.F90(251): error #8010: A pointer dummy argument with the INTENT(IN) attribute shall not appear as an pointer-object in a pointer-assignment-stmt or nullify-stmt.   [THIS]
>       var => this%variables%at(var_name)
> -------------^
> compilation aborted for /root/project/MAPL/pfio/FileMetadata.F90 (code 1)
> ```
> 
> I talked with @tclune and he had a few thoughts:
> 
> 1. `this` has to be `intent(inout)`, but I think @tclune is not the happiest with this solution
> 2. Maybe try making a temp variable to do this with, though @tclune thinks the compiler will see through it
> 3. @tclune rewrites bits of gFTL 2.
> 
> We should let @tclune comment more

I though I have fixed that. But let me try again

**Comment by tclune** [comment]
I'm going to close this.  Similar changes will be appearing organically in the evolution of MAPL3.


I tried a brut force update, and exposed some compiler bugs, so ...
